apollo-federation/src/source_aware/federated_query_graph/path_tree.rs

```diff
@@ -17,18 +22,270 @@ pub(crate) struct FederatedPathTree {
     source_entering_condition_resolutions_at_node:
         IndexMap<SelfConditionIndex, ConditionResolutionInfo>,
     condition_resolutions_at_node: IndexMap<SelfConditionIndex, ConditionResolutionInfo>,
-    childs: Vec<Arc<Child>>,
+    childs: Vec<Child>,
 }
 
-#[derive(Debug)]
+#[derive(Debug, Clone)] // XXX(@goto-bus-stop): do we want to clone this?
 pub(crate) struct Child {
     key: ChildKey,
     self_condition_resolutions_for_edge: IndexMap<SelfConditionIndex, ConditionResolutionId>,
     tree: Arc<FederatedPathTree>,
 }
 
-#[derive(Debug)]
+#[derive(Debug, Clone, PartialEq, Eq)]
 pub(crate) struct ChildKey {
     pub(crate) operation_element: Option<Arc<OperationPathElement>>,
     pub(crate) edge: Option<EdgeIndex>,
 }
+
+/// Merge two maps of conditions at a node, keeping the cheapest way to resolve each condition if
+/// there is a conflict.
+///
+/// `remapped_condition_ids` is a map of "losing" condition IDs to "winning" condition IDs in case
+/// of conflicts. This can be used to replace references to the "losing" conditions by the "winning"
+/// ones down the tree.
+fn merge_condition_resolutions(
+    condition_resolutions: &mut IndexMap<SelfConditionIndex, ConditionResolutionInfo>,
+    other_condition_resolutions: &IndexMap<SelfConditionIndex, ConditionResolutionInfo>,
+    remapped_condition_ids: &mut HashMap<ConditionResolutionId, ConditionResolutionId>,
+) {
+    for (&condition_index, resolution) in other_condition_resolutions {
+        condition_resolutions
+            .entry(condition_index)
+            .and_modify(|existing_resolution| {
+                if existing_resolution.cost > resolution.cost {
```

Comment:
We think the condition ID should always be the same. I could add a `debug_assert!()` here.